# Content
## 1. Pregatirea datelor
## 2. Selectarea algoritmului corespunzator problemei
## 3. Implementarea algoritmului si relizarea predictiei
## 4. Evaluarea modelului
## 5. Imbunatatirea modelului
## 6. Salvarea si incarcarea modelului
## 7. Combinarea tuturor etapelor


# 0. Importul bibliotecilor

In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# 1. Pregatirea datelor

Content
1. Importul datelor
2. Divizarea datelor in date feature si date labels (X si y)
3. Completarea sau excluderea datelor lipsa
4. Conversia valorilor non-numetice in numerice (feature encoding)

In [10]:
# importul datelor
heart_disease=pd.read_csv('heart-disease.csv')

In [13]:
heart_disease.head(10)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   
5   57    1   0       140   192    0        1      148      0      0.4      1   
6   56    0   1       140   294    0        0      153      0      1.3      1   
7   44    1   1       120   263    0        1      173      0      0.0      2   
8   52    1   2       172   199    1        1      162      0      0.5      2   
9   57    1   2       150   168    0        1      174      0      1.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1  
5   0     1       1  
6   0     2       1  
7   0     3       1  
8   0     3       1  
9   0     2       1

In [21]:
# Verificare datelor lipsa in coloana labels (target)
heart_disease['target'].isna().sum()

0

Deoarece nu exita nici o lipsa in coloana label se continu cu intregul set de date. Daca ar fi existat astfel de lipsuri, randul corespunzator ar fi trebuit exclus din setul de date

In [26]:
# Divizarea datelor (coloanelor) in date feature X (caracteristici) si date labels y (rezultat)
np.random.seed(9)
X=heart_disease.drop('target', axis=1)
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [27]:
y=heart_disease['target']
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

In [28]:
# Divizarea datelor in setul training si setul test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    test_size = 0.2)

In [29]:
# Verificarea dimensiunilor seturilor de training si de test
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((242, 13), (61, 13), (242,), (61,))

In [30]:
# Verificarea prezentei datelor lipsa seturile feature de training (in colona label s-a verificat mai sus)
X_train.isna().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
dtype: int64

In [31]:
# Verificarea prezentei datelor lipsa seturile feature de test (in colona label s-a verificat mai sus)
X_test.isna().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
dtype: int64

Nu exista date lipsa, deci nu este necesara procedura de completare a datelor lipsa 

In [32]:
# Verificare tipurilor de date in fiecare coloana
heart_disease.dtypes

age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca            int64
thal          int64
target        int64
dtype: object

Toate datele sunt de tip numeric deci nu este necesara procedura de conversie a datelor non-numerice in numerice

# 2a. Selectarea algoritmului corespunzator problemei

Sarcina problemei este de a determina daca o persoana este bolnava de inima sau nu, deci problema face parte din categoria problemelor de clasificare si prin urmare se va selecta un algoritm (estimator) de clasificare

Pentru selectarea algoritmului/estimator-ului corespunzator se va parcurge harta: https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html

Conform hartii scikit-learn se alege algoritmul LinearSVC

In [34]:
#Importul estimatorului LinearSVC
from sklearn.svm import LinearSVC

In [35]:
# Instantierea algoritmului LinearSVC
clf=LinearSVC()

# 3a. Implementarea algoritmului

In [39]:
#Implementarea algoritmului LinearSVC pe datele de training
clf.fit(X_train, y_train)

E:\Machine_learning\Machine_learning_projects\Classification_project\env\lib\site-packages\sklearn\svm\_base.py:985: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


LinearSVC()

# 4a. Evaluarea modelului

In [40]:
# Evaluarea modelului pe datele de test
clf.score(X_test, y_test)

0.8032786885245902

Modelul elaborat pe baza datelor de training implementand algoritmul LinearSVC are o precizie de 80% la evaluarea sa pe datele de test.
In incercarea de a creste aceasta precizie se va incerca utiliza altul algoritm conform hartii scikit-learn

# 2b. Selectarea altui algoritmului

Conform hartii scikit-learn daca nu suntem multumiti cu precizia lui modelului pe baza algoritmului LinearSVC urmatorul algorit ce poate fi utilizat este KNeighborsClassifier

In [41]:
#Importul estimatorului KNeighborsClassifier
from sklearn.neighbors import KNeighborsClassifier

In [42]:
# Instantierea algoritmului LinearSVC
clf=KNeighborsClassifier()

# 3b. Implementarea algoritmului

In [43]:
#Implementarea algoritmului KNeighborsClassifier pe datele de training
clf.fit(X_train, y_train)

KNeighborsClassifier()

# 4b. Evaluarea modelului

In [44]:
# Evaluarea modelului pe datele de test
clf.score(X_test, y_test)

0.7049180327868853

Modelul elaborat pe baza datelor de training implementand algoritmul KNeighborsClassifier are o precizie de 70% la evaluarea sa pe datele de test.
Conform hartii scikit-learn daca nu suntem multumiti de preciziile obtinute se mai incerca de a utiliza si alt algoritm 

# 2c. Selectarea altui algoritmului

Conform hartii scikit-learn daca nu suntem multumiti cu precizia lui modelului pe baza algoritmului KNeighborsClassifier urmatorul algoritm ce poate fi utilizat este unui din categoria Ensemble Classifier care de exemplu poate fi RandomForestClassifier

In [45]:
#Importul estimatorului RandomForestClassifier
from sklearn.ensemble import RandomForestClassifier

In [46]:
# Instantierea algoritmului RandomForestClassifier
clf = RandomForestClassifier()

# 3c. Implementarea algoritmului

In [47]:
#Implementarea algoritmului RandomForestClassifier pe datele de training
clf.fit(X_train, y_train)

RandomForestClassifier()

# 4c. Evaluarea modelului

In [48]:
# Evaluarea modelului pe datele de test
clf.score(X_test, y_test)

0.8032786885245902

Modelul elaborat pe baza datelor de training implementand algoritmul RandomForestClassifier are o precizie de 80% la evaluarea sa pe datele de test.
Pentru a creste aceasta precizie se poate utilizat un alt algoritm din categoria Ensemble Classifier care de exemplu poate fi AdaBoostClassifier

# 2d. Selectarea altui algoritmului

In [59]:
#Importul estimatorului AdaBoostClassifier
from sklearn.ensemble import AdaBoostClassifier

In [60]:
# Instantierea algoritmului AdaBoostClassifier
clf = AdaBoostClassifier()

# 3d. Implementarea algoritmului

In [64]:
#Implementarea algoritmului AdaBoostClassifier pe datele de training
clf.fit(X_train, y_train)

AdaBoostClassifier()

# 4d. Evaluarea modelului

In [63]:
# Evaluarea modelului pe datele de test
clf.score(X_test, y_test)

0.819672131147541

Modelul elaborat pe baza datelor de training implementand algoritmul AdaBoostClassifier are o precizie de 82% la evaluarea sa pe datele de test, cea mai mare pina acum de aceea ne vor opri aici.